src/core/lwt.ml

```diff
@@ -22,1307 +22,2746 @@
+   cancelable initial promises to cancel -- those that were created by
+   [Lwt.task]. After finding them, cancelation completes them normally with
+   [Failed Lwt.Canceled], causing an ordinary promise completion process.
+
+   To summarize, cancelation is a way to trigger an *ordinary* completion of
+   promises created with [Lwt.task], by first searching for them in the promise
+   dependency graph (which is assembled by [Lwt.bind], [Lwt.join], etc).
+
+   This backwards search is triggered only by [Lwt.cancel]. It is also possible
+   for the user to cancel a promise directly by failing it with [Lwt.Canceled],
+   but in all cases where the user can do so, the search would be redundant
+   anyway -- the user has only two ways of directly failing a promise with
+   [Lwt.Canceled] (or any exception, for that matter):
+
+   - The user can create an initial promise, then fail it through its resolver.
+     The search is redundant because it would find only the same initial promise
+     to cancel.
+   - The user can create a trivial promise by calling [Lwt.fail Lwt.Canceled].
+     The search is again redundant; in this case it would find nothing to
+     cancel.
+
+   Note that there is a quirk: only promises created by [Lwt.task] are
+   susceptible to being canceled by [Lwt.cancel], but the user can manually
+   cancel initial promises created by both [Lwt.task] and [Lwt.wait].
+
+   Due to [Lwt.cancel], promise cancelation, and therefore completion, can be
+   initiated by the user without access to a resolver. This is important for
+   reasoning about state changes in the implementation of Lwt, and is referenced
+   in some implementation detail comments.
+
+
+   6. No I/O
+
+   The Lwt core deliberately doesn't do I/O. The completion loop stops running
+   once no promises can be completed immediately. It has to be restarted later
+   by some external process.
+
+   On Unix and Windows, a separate top-level loop, typically [Lwt_main.run], is
+   necessary to repeatedly call [select], [epoll], or [kevent], and complete
+   promises that are blocked on I/O.
+
+   In JavaScript, references to promises are retained by JavaScript code, which
+   is, in turn, triggered by the JS engine. In other words, the top-level loop
+   is buried inside the JS engine.
+
+   This separation of the Lwt core from the top-level I/O loop keeps the core
+   portable.
+
+
+   7. Promise "unification"
+
+   In [Lwt.bind : 'a t -> ('a -> 'b t) -> 'b t], the outer ['b t] is created by
+   [bind] first, and returned to the user. The inner ['b t] is created by the
+   user later, and then returned to [bind]. At that point, [bind] needs to make
+   the inner and outer ['b t]s behave identically.
+
+   This is accomplished by "unifying" them, which simply means: making one of
+   the promises point to the other. One of the promises thus becomes a "proxy,"
+   and the other is its "underlying" promise.
```

Comment:
As stated in a comment in the PR thread, “unification” is confusing because it's also applies to types in ML. I prefer “proxy” and “proxying”.